File: android/src/main/java/com/reveila/android/db/PreferenceManager.java
package com.reveila.android.db;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * Java implementation of the Preference Manager.
 * Provides a robust, zero-annotation alternative to Room for simple state.
 */
public class PreferenceManager {
    private static final String PREF_NAME = "reveila_prefs";
    private static final String KEY_AGREEMENT = "user_agreement_accepted";
    private static final String KEY_TIMESTAMP = "acceptance_timestamp";
    private static final String KEY_MACHINE_ID = "acceptance_ip_or_machine_id";

    private final SharedPreferences prefs;

    public PreferenceManager(Context context) {
        this.prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
    }

    public UserPreferences getUserPreferences() {
        boolean accepted = prefs.getBoolean(KEY_AGREEMENT, false);
        
        // Handle nullable Long by checking existence
        Long timestamp = null;
        if (prefs.contains(KEY_TIMESTAMP)) {
            timestamp = prefs.getLong(KEY_TIMESTAMP, 0);
        }

        String machineId = prefs.getString(KEY_MACHINE_ID, null);

        return new UserPreferences(accepted, timestamp, machineId);
    }

    public void saveUserPreferences(UserPreferences preferences) {
        SharedPreferences.Editor editor = prefs.edit();
        
        editor.putBoolean(KEY_AGREEMENT, preferences.isUserAgreementAccepted());

        if (preferences.getAcceptanceTimestamp() != null) {
            editor.putLong(KEY_TIMESTAMP, preferences.getAcceptanceTimestamp());
        } else {
            editor.remove(KEY_TIMESTAMP);
        }

        if (preferences.getAcceptanceIpOrMachineId() != null) {
            editor.putString(KEY_MACHINE_ID, preferences.getAcceptanceIpOrMachineId());
        } else {
            editor.remove(KEY_MACHINE_ID);
        }

        editor.apply();
    }
}

File: android/src/main/kotlin/com/reveila/android/db/PreferenceManager.kt
package com.reveila.android.db

import android.content.Context
import android.content.SharedPreferences

/**
 * Modern, robust alternative to Room for simple key-value persistence.
 * Avoids annotation processing issues in complex Expo/Monorepo builds.
 */
class PreferenceManager(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("reveila_prefs", Context.MODE_PRIVATE)

    fun getUserPreferences(): UserPreferences {
        return UserPreferences(
            user_agreement_accepted = prefs.getBoolean("user_agreement_accepted", false),
            acceptance_timestamp = if (prefs.contains("acceptance_timestamp")) prefs.getLong("acceptance_timestamp", 0) else null,
            acceptance_ip_or_machine_id = prefs.getString("acceptance_ip_or_machine_id", null)
        )
    }

    fun saveUserPreferences(preferences: UserPreferences) {
        prefs.edit().apply {
            putBoolean("user_agreement_accepted", preferences.user_agreement_accepted)
            preferences.acceptance_timestamp?.let { putLong("acceptance_timestamp", it) }
            preferences.acceptance_ip_or_machine_id?.let { putString("acceptance_ip_or_machine_id", it) }
            apply()
        }
    }
}

/**
 * Keeping the data class for source compatibility.
 */
data class UserPreferences(
    val id: Int = 1,
    val user_agreement_accepted: Boolean = false,
    val acceptance_timestamp: Long? = null,
    val acceptance_ip_or_machine_id: String? = null
)
